### PR TITLE
Improve raw string parsing clarity per JEP-12

### PR DESCRIPTION
**Pull request by joshrotenberg**
This PR refactors the raw string literal parsing to use explicit character-by-character processing instead of the current string replacement approach. This makes the escape handling logic clearer and aligns the code more directly with the [JEP-12 specification](https://github.com/jmespath/jmespath.jep/blob/main/proposals/0012-raw-string-literals.md).

## Changes

- Refactor `consume_raw_string` in lexer to process characters explicitly
- Add documentation for raw string literals to rustdoc
- Add comprehensive tests for escape handling edge cases

## Behavior

No behavioral changes - this is a refactor for clarity:

- `\'` produces a literal single quote (to avoid delimiter collision)
- All other backslashes are preserved literally

## Testing

- All 861 compliance tests pass
- Added comprehensive unit tests covering 25+ escape sequence patterns
- Doc tests verify examples work correctly
- Verified identical output to previous implementation across ~20,000 generated test cases

### CODE REVIEW

File: jmespath/src/lexer.rs [hunk 1]
```diff
@@ -283,11 +283,30 @@ impl<'a> Lexer<'a> {
...
+    /// All other backslashes are preserved literally.
+    ///
+    /// See: https://github.com/jmespath/jmespath.jep/blob/main/proposals/0012-raw-string-literals.md
     #[inline]
     fn consume_raw_string(&mut self, pos: usize) -> Result<Token, JmespathError> {
-        // Note: we need to unescape here because the backslashes are passed through.
+        // Read until closing quote, then process escapes
         self.consume_inside(pos, '\'', |s| {
-            Ok(Literal(Rcvar::new(Variable::String(s.replace("\\'", "'")))))
+            // Only \' is an escape sequence - replace with literal quote
```
Comment (cetra3) [inline jmespath/src/lexer.rs:302]: How does this differ from existing behaviour?

Comment (joshrotenberg) [reply jmespath/src/lexer.rs:302]: No functional difference, just a little bit more explicit. The PR is mostly for docs/tests/clarity.

Comment (cetra3) [reply jmespath/src/lexer.rs:302]: OK what about perf difference, I am going to assume this is slower than the standard replace

Comment (joshrotenberg) [reply jmespath/src/lexer.rs:302]: You're right, it was a bit slower. Just checked in an optimization that fixes that (and seems to make it a touch faster than the original).
